src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -262,6 +262,46 @@ FabricInfo * RetrieveCurrentFabric(CommandHandler * aCommandHandler)
     return Server::GetInstance().GetFabricTable().FindFabricWithIndex(index);
 }
 
+void FailSafeCleanup(const chip::DeviceLayer::ChipDeviceEvent * event)
+{
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Call to FailSafeCleanup");
+
+    // If an AddNOC or UpdateNOC command has been successfully invoked, terminate all CASE sessions associated with the Fabric
+    // whose Fabric Index is recorded in the Fail-Safe context (see ArmFailSafe Command) by clearing any associated Secure
+    // Session Context at the Server.
+    if (event->CommissioningComplete.AddNocCommandHasBeenInvoked || event->CommissioningComplete.UpdateNocCommandHasBeenInvoked)
+    {
+        Server::GetInstance().GetSecureSessionManager().ExpireAllPairingsForFabric(event->CommissioningComplete.PeerFabricIndex);
+    }
+
+    // If an AddNOC command had been successfully invoked, achieve the equivalent effect of invoking the RemoveFabric command
+    // against the Fabric Index stored in the Fail-Safe Context for the Fabric Index that was the subject of the AddNOC
+    // command.
+    if (event->CommissioningComplete.AddNocCommandHasBeenInvoked)
+    {
+        Server::GetInstance().GetFabricTable().Delete(event->CommissioningComplete.PeerFabricIndex);
```

Comment:
This should really share more logic with the actual RemoveFabric implementation.  For example:
* This does not handle withdrawing stale operational advertisements.
* This does not close IM transactions for the fabric being removed.

We should probably have a helper method (or two, for the parts that need to happen sync and async during "normal" fabric removal) that takes a fabric index and does whatever cleanup needs to happen and is called from here, and from `FabricCleanupExchangeDelegate::OnExchangeClosing` and from `emberAfOperationalCredentialsClusterRemoveFabricCallback`.

Followup is OK for this part if desired.

Reply:
https://github.com/project-chip/connectedhomeip/issues/16412